src/libmongoc/src/mongoc/mongoc-topology-scanner.c

```diff
@@ -106,22 +106,28 @@ _jumpstart_other_acmds (mongoc_topology_scanner_node_t *node,
 static void
 _add_hello (mongoc_topology_scanner_t *ts)
 {
-   bson_t *cmd = &ts->hello_cmd;
    mongoc_server_api_t *api = ts->api;
 
+   BSON_APPEND_INT32 (&ts->hello_cmd, "hello", 1);
+   BSON_APPEND_BOOL (&ts->hello_cmd, "helloOk", true);
```

Comment:
Yes, that was the idea: use `helloOk` so that `mongoc_server_description_handle_hello` can always deduce whether `hello` is supported. Alternatively, we can change the logic to only send `helloOk` with the legacy hello command. In that case, we'd never downgrade a server after we've upgraded it to `hello`, but in that case the first failing `hello` would trip the error condition making the next heartbeat use the handshake command which defaults to using legacy hello.

Not downgrading servers would also take care of some other issues you mentioned below.

Reply:
We still always set `helloOk: true` but it's no longer reset when reading the server description. This was done so that we still set `hello_ok` on the server description when having an API version declared (in which case we assemble the handshake command from `hello_cmd` and never set `hello_ok` manually).

Reply:
I do not understand the last comment, but the implementation looks correct. `_build_handshake_cmd` does not inspect the node's `hello_ok` field. But it does inspect the scanner's `api` field to determine if the handshake should be a `hello` or legacy hello.

The node's `hello_ok` field is updated from the server description. But if an API version is set, the `hello_ok` field is effectively ignored. Both the handshake and monitoring command appear to correctly return the `hello` command.